cluster-autoscaler/cloudprovider/aws/aws_wrapper.go

```diff
@@ -233,32 +235,455 @@ func (m *awsWrapper) getInstanceTypeByLaunchTemplate(launchTemplate *launchTempl
 		LaunchTemplateName: aws.String(launchTemplate.name),
 		Versions:           []*string{aws.String(launchTemplate.version)},
 	}
-
 	start := time.Now()
 	describeData, err := m.DescribeLaunchTemplateVersions(params)
 	observeAWSRequest("DescribeLaunchTemplateVersions", err, start)
 	if err != nil {
 		return "", err
 	}
+	if len(describeData.LaunchTemplateVersions) == 0 {
+		return "", fmt.Errorf("unable to find versions for launch template %s", launchTemplate.name)
+	}
+
+	launchTemplateData := describeData.LaunchTemplateVersions[0].LaunchTemplateData
+	if launchTemplateData == nil {
+		return "", fmt.Errorf("no data found for launch template %s, version %s", launchTemplate.name, launchTemplate.version)
+	}
+
+	instanceType := ""
+	if launchTemplateData.InstanceType != nil {
+		instanceType = *launchTemplateData.InstanceType
+	} else if launchTemplateData.InstanceRequirements != nil && launchTemplateData.ImageId != nil {
+		requirementsRequest, err := m.getRequirementsRequestFromEC2(launchTemplateData.InstanceRequirements)
+		if err != nil {
+			return "", fmt.Errorf("unable to get instance requirements request")
+		}
+		instanceType, err = m.getInstanceTypeFromInstanceRequirements(*launchTemplateData.ImageId, requirementsRequest)
+		if err != nil {
+			return "", err
+		}
+	}
+	if len(instanceType) == 0 {
+		return "", fmt.Errorf("unable to find instance type using launch template")
+	}
+
+	return instanceType, nil
+}
+
+func (m *awsWrapper) getInstanceTypeFromRequirementsOverrides(policy *mixedInstancesPolicy) (string, error) {
+	if policy.launchTemplate == nil {
+		return "", fmt.Errorf("no launch template found for mixed instances policy")
+	}
 
+	describeTemplateInput := &ec2.DescribeLaunchTemplateVersionsInput{
+		LaunchTemplateName: aws.String(policy.launchTemplate.name),
+		Versions:           []*string{aws.String(policy.launchTemplate.version)},
+	}
+
+	start := time.Now()
+	describeData, err := m.DescribeLaunchTemplateVersions(describeTemplateInput)
+	observeAWSRequest("DescribeLaunchTemplateVersions", err, start)
+	if err != nil {
+		return "", err
+	}
 	if len(describeData.LaunchTemplateVersions) == 0 {
 		return "", fmt.Errorf("unable to find template versions")
 	}
```

Comment:
you need to place the same nil-guard on `lt.LaunchTemplateData` that you did above on new line 249. You might consider DRY'ing this up into a helper method since you are calling `m.DescribeLaunchTemplateVersions` numerous times in both `aws_wrapper.go` and `aws_manager.go`

Reply:
Makes sense! Will do 👍